core/src/main/java/org/infinispan/util/concurrent/locks/impl/InfinispanLock.java

```diff
@@ -412,13 +418,13 @@ public InvocationStage toInvocationStage(Supplier<TimeoutException> timeoutSuppl
             return checkState(notifier.getNow(lockState), InvocationStage::completedNullStage,
                   ExceptionSyncInvocationStage::new, timeoutSupplier);
          }
-         return new SimpleAsyncInvocationStage(notifier.thenApplyAsync(state -> {
+         return new SimpleAsyncInvocationStage(CompletionStages.continueOnExecutor(notifier.thenApplyAsync(state -> {
             Object rv = checkState(state, () -> null, throwable -> throwable, timeoutSupplier);
             if (rv != null) {
                throw (RuntimeException) rv;
             }
             return null;
-         }, executor));
+         }, blockingExecutor), nonBlockingExecutor, this));
```

Comment:
This should be in a separate commit for ISPN-11295.

Also, I would expect no blocking in the vast majority of cases, so we need an optimized path for that case and/or a PerfAck test to see if the extra thread switching affects performance.

Reply:
In vast majority of cases this code is never even hit, because the notifier is normally usually done above. So I would think the thread switching here is probably not going to be an issue, and actually has a very high likely hood of actually blocking.

The reason the change is needed is because it then resumed the operation on the blocking thread, which we don't want.

Reply:
I can make a separate commit for it.

Reply:
I will take a closer look at this in general with this pass though. I should be able to solve it how I would want for ISPN-11295.

Reply:
> In vast majority of cases this code is never even hit, because the notifier is normally usually done above. 

That depends on the access access pattern. PerfAck tests make each stressor use separate keys, so there is no contention, but when there is lots of contention the stage wouldn't be done.

The new approach resuming directly on a non-blocking thread looks much better!

Reply:
Yeah, I am not sure why it needed a blocking thread before. So I am glad to change it to this now.